src/attention.jl

```diff
@@ -0,0 +1,125 @@
+- `value`: Value array of size `(v_dim, kv_len, batch_size...)`.
+- `bias`: Either `nothing` or an input array broadcastable to size `(kv_len, q_len, num_heads, batch_size)`. 
+          Can also be set to `mask=:causal` to apply a causal mask. Default `nothing`.
+- `dropout_fn`: A dropout function to apply on the attention scores. Default `nothing`. 
+- `mask`: Either `nothing` or an input array broadcastable to size `(kv_len, q_len, num_heads, batch_size)`. 
+          Can also be set to `mask=:causal` to apply a causal mask. Default `nothing`.
+- `num_heads`: Number of heads to split the input arrays into. Default `1`.
+
+# Examples
+    
+```julia
+q, k, v = rand(10, 20, 2), rand(10, 30, 2), rand(20, 30, 2)
+y, α = dot_product_attention(q, k, v)
+```
+"""
+function dot_product_attention(q::AA{N}, k::AA{N}, v::AA{N}; num_heads=1, kws...) where N
+    batch_size = size(q)[3:end]
+    
+    batch_size == size(k)[3:end] == size(v)[3:end] || throw(ArgumentError("Batch dimensions have to be the same."))
+    size(q, 1) == size(k, 1) || throw(ArgumentError("First dimension in query and key has to be the same."))
+    size(k, 2) == size(v, 2)  || throw(ArgumentError("Second dimension in key and value has to be the same."))
+    
+    q, k, v = map(x -> reshape(x, size(x, 1), size(x, 2), :), (q, k, v))
+
+    # Multihead attention. TODO create fastpath for singlehead attention.
+    q, k, v = split_heads.((q, k, v), num_heads)
+    x, α = _dot_product_attention(q, k, v; kws...)
+    x = join_heads(x)
+
+    x = reshape(x, size(x, 1), size(x, 2), batch_size...)
+    α = reshape(α, size(α)[1:3]..., batch_size...)
+    return x, α
+end
+
+function _dot_product_attention(q::AA4, k::AA4, v::AA4; 
+        dropout_fn=nothing, bias=nothing, mask=nothing)
+
+    α = dot_product_attention_scores(q, k; dropout_fn, bias, mask)
+    # [α] = [kv_len, q_len, num_heads, batch_size]
+    
+    # The following permutedims and batched_mul are equivalent to
+    # @tullio x[d, h, i, b] := α[j, i, h, b] * v[d, h, j, b]
+    vt = permutedims(v, (1, 3, 2, 4))
+    x = batched_mul(vt, α)
+    x = permutedims(x, (1, 3, 2, 4))
+    # [x] = [kv_dim ÷ num_heads, num_heads, q_len, batch_size]
+    return x, α
+end
+
+"""
+    dot_product_attention_scores(query, key; [bias, droput_fn, mask])
+
+Return the attention scores for the [`dot_product_attention`](@ref).
+
+Input arrays must have dimensions `(num_features ÷ num_heads, num_heads, sequence_length, batch_size)`
+
+"""
+function dot_product_attention_scores(q::AA4{T}, k::AA4{T}; 
+            dropout_fn=nothing, mask=nothing, bias=nothing) where T
+
+    q  = q ./ √T(size(q, 1))
+    
+    # The following permutedims and batched_mul are equivalent to
+    # @tullio α[j, i, h, b] := q[d, h, i, b] * k[d, h, j, b]
+    kt = permutedims(k, (3, 1, 2, 4))
+    qt = permutedims(q, (1, 3, 2, 4))
+    α = batched_mul(kt, qt)
+    # [α] = [kv_len, q_len, num_heads, batch_size]
+
+    if bias !== nothing
+        α = α .+ bias
```

Comment:
yes, bias is used for score reshaping based on locality and on context features. 
Can also be used for masking, but it is more intuitive and convenient to pass a boolean mask instead of a array with infs.

Reply:
If it must exist then  it probably ought to be a positional argument. AD with keywords is asking for pain.